mm2src/mm2_net/src/wasm/body_stream.rs

```diff
@@ -0,0 +1,411 @@
+            self.decode_base64_chunk()?;
+        } else {
+            self.buf.put(bytes)
+        }
+
+        Ok(())
+    }
+
+    fn take(&mut self, length: usize) -> BytesMut {
+        let new_buf = self.buf.split_off(length);
+        std::mem::replace(&mut self.buf, new_buf)
+    }
+}
+
+impl Deref for EncodedBytes {
+    type Target = BytesMut;
+
+    fn deref(&self) -> &Self::Target { &self.buf }
+}
+
+impl DerefMut for EncodedBytes {
+    fn deref_mut(&mut self) -> &mut Self::Target { &mut self.buf }
+}
+
+/// Represents the state of reading the response body, including compression flags, data lengths, trailers, and the done state.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ReadState {
+    CompressionFlag,
+    DataLength,
+    Data(u32),
+    TrailerLength,
+    Trailer(u32),
+    Done,
+}
+
+impl ReadState {
+    fn is_done(&self) -> bool { matches!(self, ReadState::Done) }
+
+    fn finished_data(&self) -> bool {
+        matches!(self, ReadState::TrailerLength)
+            || matches!(self, ReadState::Trailer(_))
+            || matches!(self, ReadState::Done)
+    }
+}
+
+/// Handles the HTTP response body, decoding data, and extracting trailers
+#[pin_project]
+pub struct ResponseBody {
+    #[pin]
+    body_stream: BodyStream,
+    buf: EncodedBytes,
+    incomplete_data: BytesMut,
+    data: Option<BytesMut>,
+    trailer: Option<HeaderMap>,
+    state: ReadState,
+    finished_stream: bool,
+}
+
+impl ResponseBody {
+    /// Creates a new `ResponseBody` based on a ReadableStream and content type.
+    pub(crate) async fn new(body_stream: ReadableStream, content_type: &str) -> Result<Self, PostGrpcWebErr> {
+        let body_stream: ReadableStreamDefaultReader = body_stream
+            .get_reader()
+            .dyn_into()
+            .map_err(|err| PostGrpcWebErr::BadResponse(format!("{err:?}")))?;
+
+        Ok(Self {
+            body_stream: BodyStream::new(body_stream).await?,
+            buf: EncodedBytes::new(content_type)?,
+            incomplete_data: BytesMut::new(),
+            data: None,
+            trailer: None,
+            state: ReadState::CompressionFlag,
+            finished_stream: false,
+        })
+    }
+
+    fn read_stream(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), PostGrpcWebErr>> {
+        if self.finished_stream {
+            return Poll::Ready(Ok(()));
+        }
+
+        let this = self.project();
+
+        match ready!(this.body_stream.poll_data(cx)) {
+            Some(Ok(data)) => {
+                if let Err(e) = this.buf.append(data) {
+                    return Poll::Ready(Err(e));
+                }
+
+                Poll::Ready(Ok(()))
+            },
+            Some(Err(e)) => Poll::Ready(Err(e)),
+            None => {
+                *this.finished_stream = true;
+                Poll::Ready(Ok(()))
+            },
+        }
+    }
+
+    fn step(self: Pin<&mut Self>) -> Result<(), PostGrpcWebErr> {
```

Comment:
I believe the code of this function can be made better, I will show a few examples in the below comments

Reply:
done